include/envoy/http/codec.h

```diff
@@ -148,11 +148,32 @@ class ConnectionCallbacks {
 };
 
 /**
- * A list of options that can be specified when creating a codec.
+ * HTTP/2 codec settings
  */
-class CodecOptions {
-public:
-  static const uint64_t NoCompression = 0x1;
+struct Http2Settings {
+  uint32_t hpack_table_size_{DEFAULT_HPACK_TABLE_SIZE};
+  uint32_t max_concurrent_streams_{DEFAULT_MAX_CONCURRENT_STREAMS};
+  uint32_t initial_window_size_{DEFAULT_INITIAL_WINDOW_SIZE};
+
+  // disable HPACK compression
+  static const uint32_t MIN_HPACK_TABLE_SIZE = 0;
+  // initial value from HTTP/2 spec, same as NGHTTP2_DEFAULT_HEADER_TABLE_SIZE from nghttp2
+  static const uint32_t DEFAULT_HPACK_TABLE_SIZE = 4 * 1024;
+  // a 16MiB maximum, hope it's more than enough
+  static const uint32_t MAX_HPACK_TABLE_SIZE = 16 * 1024 * 1024;
+
+  static const uint32_t MIN_MAX_CONCURRENT_STREAMS = 1;
+  static const uint32_t DEFAULT_MAX_CONCURRENT_STREAMS = 1024;
+  // All streams are 2^31. Client/Server streams are half that. Just to be on the safe
+  // side we do 2^29. Same as MAX_STREAMS in source/common/http/http2/conn_pool.h.
+  static const uint32_t MAX_MAX_CONCURRENT_STREAMS = 1 << 29;
+
+  // initial value from HTTP/2 spec, same as NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE.
+  // We only support increasing window size now, so this is also the minimum.
+  static const uint32_t MIN_INITIAL_WINDOW_SIZE = (1 << 16) - 1;
+  static const uint32_t DEFAULT_INITIAL_WINDOW_SIZE = 256 * 1024 * 1024;
```

Comment:
256MiB is from the old code here : https://github.com/lyft/envoy/blob/a8c446aafa2af08f525dbab2b052f25834b78139/source/common/http/http2/codec_impl.h#L220 

it's used for both connection-level and stream-level window.

i added a new initial_connection_window_size_ field to seperate the two window.